src/Engine/IntegrityCheckTrait.php

```diff
@@ -172,4 +165,33 @@ protected function checkDataTypeCompatible(ContextDefinitionInterface $context_d
     }
   }
 
+  /**
+   * Adds provided variables to the execution metadata state.
+   *
+   * @param CoreContextAwarePluginInterface $plugin
+   *   The action or condition plugin that may provide variables.
+   * @param \Drupal\rules\Engine\ExecutionMetadataStateInterface $metadata_state
+   *   The excution metadata state to add variables to.
+   */
+  public function doPrepareExecutionMetadataState(CoreContextAwarePluginInterface $plugin, ExecutionMetadataStateInterface $metadata_state) {
```

Comment:
doPrepareExecutionMetadataState() is a bad name imo as what preparing means differs by the plugin. but the trait only does "adding variables to it", so the name should be about that. e.g. addProvidedVariablesToExecutionMetadataState :)